File: app/src/main/java/com/alexkn/syntact/presentation/createbucket/CreateBucketFragment.java
package com.alexkn.syntact.presentation.createbucket;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.LinearSnapHelper;
import androidx.recyclerview.widget.RecyclerView;

import com.alexkn.syntact.R;
import com.alexkn.syntact.app.ApplicationComponentProvider;
import com.alexkn.syntact.restservice.Template;

import java.util.List;
import java.util.Locale;

public class CreateBucketFragment extends Fragment {

    private static final String TAG = CreateBucketFragment.class.getSimpleName();

    private CreateBucketViewModel viewModel;

    private View addButton;

    private Locale selectedLanguage;

    private Template selectedTemplate;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.bucket_create_fragment, container, false);
        viewModel = ViewModelProviders
                .of(this, ((ApplicationComponentProvider) getActivity().getApplication()).getApplicationComponent().createBucketViewModelFactory())
                .get(CreateBucketViewModel.class);

        addButton = view.findViewById(R.id.addBucketButton);
        EditText words = view.findViewById(R.id.editText);

        List<Locale> dataset = viewModel.getAvailableBuckets();
        dataset.remove(Locale.getDefault());
        selectedLanguage = dataset.get(0);

        RecyclerView languageRecyclerView = view.findViewById(R.id.selectLanguageRecyclerView);
        LinearLayoutManager languageLayoutManager = new LinearLayoutManager(getContext(), RecyclerView.HORIZONTAL, false);
        languageRecyclerView.setLayoutManager(languageLayoutManager);
        ChooseLanguageAdapter adapter = new ChooseLanguageAdapter(dataset);
        languageRecyclerView.setAdapter(adapter);
        LinearSnapHelper linearSnapHelper = new LinearSnapHelper();
        linearSnapHelper.attachToRecyclerView(languageRecyclerView);
        languageRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {

            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {

                if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                    int itemPosition = languageLayoutManager.findFirstVisibleItemPosition();
                    if (itemPosition != RecyclerView.NO_POSITION) {
                        selectedLanguage = adapter.getItemAt(itemPosition);
                    }
                }
            }
        });

        RecyclerView templateRecyclerView = view.findViewById(R.id.selectTemplateRecyclerView);
        LinearLayoutManager templateLayoutManager = new LinearLayoutManager(getContext(), RecyclerView.HORIZONTAL, false);
        templateRecyclerView.setLayoutManager(templateLayoutManager);

        ChooseTemplateAdapter chooseTemplateAdapter = new ChooseTemplateAdapter();
        templateRecyclerView.setAdapter(chooseTemplateAdapter);
        LinearSnapHelper linearSnapHelper1 = new LinearSnapHelper();
        linearSnapHelper1.attachToRecyclerView(templateRecyclerView);

        templateRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {

            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {

                if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                    int itemPosition = templateLayoutManager.findFirstVisibleItemPosition();
                    if (itemPosition != RecyclerView.NO_POSITION) {
                        selectedTemplate = chooseTemplateAdapter.getItemAt(itemPosition);
                    }
                }
            }
        });

        addButton.setOnClickListener(v -> {

            viewModel.addBucket(selectedLanguage, selectedTemplate, words.getText().toString());//TODO
            Navigation.findNavController(view).popBackStack();
        });

        viewModel.getAvailableTemplates().observe(getViewLifecycleOwner(), chooseTemplateAdapter::setData);

        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {

        super.onActivityCreated(savedInstanceState);
        viewModel = ViewModelProviders.of(this).get(CreateBucketViewModel.class);
    }
}

File: app/src/main/java/com/alexkn/syntact/presentation/createbucket/CreateBucketFragment.kt
package com.alexkn.syntact.presentation.createbucket

import android.os.Bundle
import android.os.Handler
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.Navigation
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.LinearSnapHelper
import androidx.recyclerview.widget.RecyclerView
import com.alexkn.syntact.R
import com.alexkn.syntact.app.ApplicationComponentProvider
import com.alexkn.syntact.data.model.Template
import com.google.android.material.bottomsheet.BottomSheetBehavior
import kotlinx.android.synthetic.main.create_bucket_fragment.*
import kotlinx.android.synthetic.main.create_bucket_language_sheet.*
import java.util.*
import java.util.function.Consumer


class CreateBucketFragment : Fragment() {

    private lateinit var viewModel: CreateBucketViewModel

    private var selectedLanguage: Locale? = null

    private var filteredTemplates: MediatorLiveData<List<Template>> = MediatorLiveData()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.create_bucket_fragment, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {

        viewModel = ViewModelProvider(this, (activity!!.application as ApplicationComponentProvider).applicationComponent.createBucketViewModelFactory())
                .get(CreateBucketViewModel::class.java)

        backButton.setOnClickListener { Navigation.findNavController(it).popBackStack() }

        val languageLayoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
        languageRecyclerView.layoutManager = languageLayoutManager
        val chooseLanguageAdapter = ChooseLanguageAdapter(viewModel.availableLanguages)
        languageRecyclerView.adapter = chooseLanguageAdapter

        val bottomSheetBehavior = BottomSheetBehavior.from(languageSheet)

        chooseLanguageAdapter.getLanguage().observe(viewLifecycleOwner, Observer { locale ->
            selectedLanguage = locale
            val resId = context!!.resources.getIdentifier(locale.language, "drawable", context!!.packageName)
            flagImage.setImageResource(resId)
            Handler().postDelayed({ bottomSheetBehavior.state = BottomSheetBehavior.STATE_COLLAPSED }, 300)
        })

        val templateLayoutManager = LinearLayoutManager(context, RecyclerView.HORIZONTAL, false)
        selectTemplateRecyclerView.layoutManager = templateLayoutManager

        val chooseTemplateAdapter = ChooseTemplateAdapter()
        selectTemplateRecyclerView.adapter = chooseTemplateAdapter
        val linearSnapHelper = LinearSnapHelper()
        linearSnapHelper.attachToRecyclerView(selectTemplateRecyclerView)
        selectTemplateRecyclerView.visibility = View.GONE

        chooseLanguageSheetLabel.text = "0 Templates"
        bottomSheetBehavior.setBottomSheetCallback(object : BottomSheetBehavior.BottomSheetCallback() {

            override fun onSlide(bottomSheet: View, slideOffset: Float) {
                arrowUp.rotation = slideOffset * -180
                flagImage.alpha = (1 - slideOffset)
            }

            override fun onStateChanged(bottomSheet: View, newState: Int) = Unit
        })
        bottomSheetBehavior.isHideable = false

        flagImage.clipToOutline = true

        chooseTemplateAdapter.createListener = Consumer {
            viewModel.addBucketFromExistingTemplate(it)
            Navigation.findNavController(view).popBackStack()
        }

        filteredTemplates.addSource(viewModel.availableTemplates) {
            filteredTemplates.value = it.filter { t -> selectedLanguage?.let { t.language == selectedLanguage } ?: true }
        }

        filteredTemplates.addSource(chooseLanguageAdapter.getLanguage()) {
            val list = viewModel.availableTemplates.value
            filteredTemplates.value = list?.filter { t -> t.language == it }
        }

        filteredTemplates.observe(viewLifecycleOwner, Observer {
            chooseTemplateAdapter.submitList(it)
            progressBar2.visibility = View.GONE
            selectTemplateRecyclerView.visibility = View.VISIBLE
            when {
                it.size == 1 -> chooseLanguageSheetLabel.text = "1 Template"
                else -> chooseLanguageSheetLabel.text = "%d Templates".format(it.size)
            }
        })
    }

}


